text/0000-stabilize-marker-freeze.md

```diff
@@ -79,21 +79,22 @@ Note that `T: Freeze` is a shallow property: `T` is still allowed to contain int
 provided that it is behind an indirection (such as `Box<UnsafeCell<U>>`).
 Notable `!Freeze` types are [`UnsafeCell`](core::cell::UnsafeCell) and its safe wrappers
 such as the types in the [`cell` module](core::cell), [`Mutex`](std::sync::Mutex), and [atomics](core::sync::atomic).
-Any type which contains any one of these without indirection is also `!Freeze`.
+Any type which contains a `!Freeze` type without indirection is also `!Freeze`.
```

Comment:
Is this true?  I would expect a type that contains a `!Freeze` type to be `?Freeze`, and that seems to be what `rustc` expects also:

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&code=%23%21%5Bfeature%28freeze%2C+freeze_impls%29%5D%0A%23%21%5Bfeature%28negative_impls%29%5D%0A%23%21%5Ballow%28internal_features%29%5D%0A%0Ause+core%3A%3A%7Bcell%3A%3AUnsafeCell%2C+marker%3A%3AFreeze%7D%3B%0A%0Astruct+S%28UnsafeCell%3C%28%29%3E%29%3B%0A%2F%2Fimpl+%21Freeze+for+S+%7B%7D+%2F%2F+Uncomment+to+make+%60S%3A+%21Freeze%60.%0A%0Atrait+Tr%3CT%3E+%7B%7D%0A%0Aimpl+Tr%3CS%3E+for+%28%29+%7B%7D%0Aimpl%3CT%3A+Freeze%3E+Tr%3CT%3E+for+%28%29+%7B%7D%0A%2F%2F~%5E+ERROR+conflicting+implementations+of+trait+%60Tr%3CS%3E%60+for+type+%60%28%29%60%0A%0Afn+main%28%29+%7B%7D%0A)

If this should be changed, it'd be worth checking elsewhere in the document for other cases like this.

Reply:
```suggestion
Any type which contains a non-`Freeze` type without indirection also does not implement `Freeze`.
```

Yeah, if negative impls are part of the reasoning, then "a type that does not implement `Freeze`"/"a non-`Freeze` type" does not mean the same thing as "a `!Freeze` type", and I think the intent of this section is that the container type is the former. The contained type could be either equivalently though, I think[^1].

Also, `Mutex` is described above as a "notable `!Freeze` type", though there is currently no explicit negative impl for it (`Mutex` is "a type not implementing `Freeze`", but not "a type implementing `!Freeze`" according to negative-impls logic). Perhaps that section could instead say "Notable non-`Freeze` types are [...]"

[^1]: since `Freeze` is an auto trait, so any "non-`Freeze` type" either has an explicit `impl !Freeze`, or contains such a type.

-----

Note that using `Send` or `Sync` in place of `Freeze` in the playground snippet gives the same error, but `Clone` does not, which shows that this is probably `auto`-trait-specific logic somehow.

Reply:
@zachs18 thanks for clearing up the section. Indeed me using `!Freeze` as a shortcut for non-`Freeze` may have been misleading :)

One could argue that `UnsafeCell` and atomics _should_ be `!Freeze`, but that `Mutex` could be simply non-`Freeze` by virtue of the data it contains being inlined being an implementation detail